dart-lang/native
# [ffigen] Method filtering

Implement method filtering for ObjC interfaces and protocols:

The YAML schema is essentially copying the existing include/exclude keys, and nesting them below a regex key. Example from the test:

```yaml
objc-interfaces:
  include:
    - MethodFilteringTestInterface
  member-filter:
    Metho.*ngTe.*rface: # The full match rule below takes precedence.
      include:
        - .*
    MethodFilteringTestInterface:
      include:
        - includedStaticMethod
        - excludedStaticMethod   # The exclude rule takes precedence.
        - inc.*Ins.*Me.*od:wi.*  # Methods use original names with : delimiters.
        - includedProperty
      exclude:
        - excluded.*
objc-protocols:
  include:
    - MethodFilteringTestProtocol
  member-filter:
    Met.*ringT.*ocol:
      include:
        - includedProtocolMethod
      # If include is non-empty, everything else is excluded.
```

The behavior is a combination of the existing includer logic and the member renaming logic:

- The filter regexes select which interface/protocol to apply the include/exclude rules to, and are separated into "full matches" and regex patterns.
- Full matchers are basically just any string that doesn't have regex symbols in it (this logic is copied directly from the member renaming code).
- If an interface matches a regex pattern and a full match, the full match takes precedence (if it matches multiple regexes, the first one wins).
- If a method matches both an include rule and an exclude rule, the exclude rule wins.
- If any include rules are specified for an interface, its methods default to excluded by default. Otherwise it defaults to included (to match existing behavior).
- All match rules see the full original names of the interfaces and methods. In particular, method names are presented in the ObjC format, with `:` separating the argument names.

Implementation:
- Add config infrastructure for member filtering.
  - Added `shouldIncludeMember` method to `DeclarationFilters`
  - The yaml based implementation lives in `YamlMemberIncluder`
  - Parsing occurs in `declarationConfigExtractor`, and pretty much just repurposes the existing parsing logic for includers.
- Apply method filters in the interface and protocol parsers.

Other changes:
- I realized we had the infra for member renaming already, but weren't using it for interfaces, so I added that. We already had the config plumbing for it, so I just had to add an `originalName`/`name` distinction to `ObjCMethod`, apply the renaming in the parsers, and use the `name` instead of the `originalName` when generating the method name for codegen.
- Small refactor to `YamlIncluder` to share logic between the include part and the exclude part. That logic now lives in the `YamlFilter` class.
- Wherever the readme talks about nested keys in the YAML schema, I've changed them from `outer->inner` to `outer.inner`, as this is more idiomatic.

Fixes #251
Fixes #1269
